### PR TITLE
extract activity schema

### PR DESCRIPTION
**Pull request by olinux**

### CODE REVIEW

File: schemas/research/protocolExecution.schema.tpl.json [hunk 1]
```diff
@@ -1,56 +1,7 @@
 {
   "_type": "https://openminds.ebrains.eu/core/ProtocolExecution",
-  "required": [
-    "input", 
-    "output", 
-    "protocol"
```
Comment (lzehl) [inline schemas/research/protocolExecution.schema.tpl.json:6]: `protocol` still needs to be required in addition to `input` and `output` defined in `Activity`.

File: schemas/research/activity.schema.tpl.json [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{
+  "required": [
+    "input", 
+    "output"
+  ],
+  "properties": {
+    "description": {
+      "type": "string",
+      "_instruction": "Enter a description of this protocol execution."
```
Comment (lzehl) [inline schemas/research/activity.schema.tpl.json]: _instruction need to be generalized in the concept schema. (e.g., replace "protocol execution" with "activity")

File: schemas/research/activity.schema.tpl.json [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    },
+    "input": {
+      "type": "array",
+      "minItems": 1,
+      "uniqueItems": true,
+      "_instruction": "Add all inputs (subject state, tissue sample state, file instance or file bundle) used by this protocol execution.",
```
Comment (lzehl) [inline schemas/research/activity.schema.tpl.json]: _instruction need to be generalized in the concept schema. (e.g., replace "protocol execution" with "activity")
+ if I'm not mistaken, `_linkedTypes` will change for `input` in extensions, therefore the list in the brackets should be removed

File: schemas/research/activity.schema.tpl.json [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    },
+    "lookupLabel": {
+      "type": "string",
+      "_instruction": "Enter a lookup label for this protocol execution that may help you to more easily find it again."
+    },
+    "output": {
+      "type": "array",
+      "minItems": 1,
+      "uniqueItems": true,
+      "_instruction": "Add all outputs (subject state, tissue sample state, file instance or file bundle) generated by this protocol execution.",
```
Comment (lzehl) [inline schemas/research/activity.schema.tpl.json]: _instruction need to be generalized in the concept schema. (e.g., replace "protocol execution" with "activity")
+ if I'm not mistaken, `_linkedTypes` will change for `output` in extensions, therefore the list in the brackets should be removed

File: schemas/research/activity.schema.tpl.json [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        "https://openminds.ebrains.eu/core/File",
+        "https://openminds.ebrains.eu/core/SubjectGroupState",
+        "https://openminds.ebrains.eu/core/SubjectState",
+        "https://openminds.ebrains.eu/core/TissueSampleCollectionState",
+        "https://openminds.ebrains.eu/core/TissueSampleState"
+      ]
+    },
+    "lookupLabel": {
+      "type": "string",
+      "_instruction": "Enter a lookup label for this protocol execution that may help you to more easily find it again."
```
Comment (lzehl) [inline schemas/research/activity.schema.tpl.json]: _instruction need to be generalized in the concept schema. (e.g., replace "protocol execution" with "activity")

File: schemas/research/activity.schema.tpl.json [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        "https://openminds.ebrains.eu/core/SubjectState",
+        "https://openminds.ebrains.eu/core/TissueSampleCollectionState",
+        "https://openminds.ebrains.eu/core/TissueSampleState"
+      ]
+    },
+    "parameterSet": {
+      "type": "array",
+      "minItems": 1,
+      "uniqueItems": true,
+      "_instruction": "Add all important parameters grouped in context-specific sets that were used in this protocol execution.",
```
Comment (lzehl) [inline schemas/research/activity.schema.tpl.json]: _instruction need to be generalized in the concept schema. (e.g., replace "protocol execution" with "activity")

File: schemas/research/activity.schema.tpl.json [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      "_embeddedTypes": [
+         "https://openminds.ebrains.eu/core/ParameterSet"
+      ]
+    },
+    "semanticallyAnchoredTo": {
+      "type": "array",
+      "minItems": 1,
+      "uniqueItems": true,
+      "_instruction": "Add all anatomical entities to which the outputs of this protocol execution can be semantically anchored to.",
```
Comment (lzehl) [inline schemas/research/activity.schema.tpl.json]: _instruction need to be generalized in the concept schema. (e.g., replace "protocol execution" with "activity")

File: schemas/research/activity.schema.tpl.json [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+      "_linkedTypes": [
+         "https://openminds.ebrains.eu/sands/AnatomicalEntity"
+      ]
+    },
+    "studyTarget": {
+      "type": "array",
+      "minItems": 1,
+      "uniqueItems": true,
+      "_instruction": "Add all study targets of this model version.",
```
Comment (lzehl) [inline schemas/research/activity.schema.tpl.json]: _instruction need to be generalized in the concept schema. (e.g., replace "protocol execution" with "activity")

**Review by lzehl — CHANGES_REQUESTED** [review]
General comment: we need to decide if the "_instruction" should be generalized in a concept schema or if we want the "_instruction" to be always precise and therefore always be provided in the extension (actual schema) which would mean all properties need to be listed again... Or if we can avoid the "naming" of the actual schema in these instructions to make them precise for the property but generic for the schema. Let me know what you think and I'm happy to help implementing the decision we make.

**Comment by lzehl** [comment]
@apdavison we need your help here. Should the Activity hold all the properties that are currently in there or should it only be "input" and "output" ??

File: schemas/research/activity.schema.tpl.json [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    },
+    "input": {
+      "type": "array",
+      "minItems": 1,
+      "uniqueItems": true,
+      "_instruction": "Add all inputs used by this activity.",
+      "_linkedTypes": [
+        "https://openminds.ebrains.eu/core/FileBundle",
+        "https://openminds.ebrains.eu/core/File",
+        "https://openminds.ebrains.eu/core/SubjectGroupState",
```
Comment (apdavison) [inline schemas/research/activity.schema.tpl.json]: I think `Activity` should have only `File` and `FileBundle` as input and output, the others should be added in the `input`/`output` fields of `ProtocolExecution`.

File: schemas/research/activity.schema.tpl.json [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        "https://openminds.ebrains.eu/core/SubjectState",
+        "https://openminds.ebrains.eu/core/TissueSampleCollectionState",
+        "https://openminds.ebrains.eu/core/TissueSampleState"
+      ]
+    },
+    "lookupLabel": {
```
Comment (apdavison) [inline schemas/research/activity.schema.tpl.json:29]: There is a widely used term for "lookup label": "tag" !

File: schemas/research/activity.schema.tpl.json [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+      "_instruction": "Add all important parameters grouped in context-specific sets that were used in this activity.",
+      "_embeddedTypes": [
+         "https://openminds.ebrains.eu/core/ParameterSet"
+      ]
+    },
+    "semanticallyAnchoredTo": {
+      "type": "array",
+      "minItems": 1,
+      "uniqueItems": true,
+      "_instruction": "Add all anatomical entities to which the outputs of this activity can be semantically anchored to.",
```
Comment (apdavison) [inline schemas/research/activity.schema.tpl.json]: two "to"s is too many :-) Either "to which...anchored." or "which...anchored to". I prefer the former.

**Review by apdavison — COMMENTED** [review]
looks good in general, a couple of minor things, and a discussion needed about what to put in input/output

File: schemas/research/activity.schema.tpl.json (hunk 9 above)
Comment (lzehl) [reply schemas/research/activity.schema.tpl.json]: I agree

File: schemas/research/activity.schema.tpl.json (hunk 10 above)
Comment (lzehl) [reply schemas/research/activity.schema.tpl.json:29]: We introduced lookup label now in many schemas. I would not like to change it again, and at least in the way it is used in the KG it is not a "tag" (for me at least). It is only meant for curators (or users) if they set up a metadata graph for a research product to be able to quicker find the entities again (we actually did not want to introduce it at all, but working purely with the UUIDs turned out to be too cumbersome). Therefore the lookup labels will never be visible in the KG Search. A "tag" could be misunderstood as a general tag (like a keyword) that might also appear in the Search later.

File: schemas/research/activity.schema.tpl.json (hunk 11 above)
Comment (lzehl) [reply schemas/research/activity.schema.tpl.json]: You are right... we'll gonna correct this.

**Comment by lzehl** [comment]
> looks good in general, a couple of minor things, and a discussion needed about what to put in input/output

@apdavison just to be sure: so besides input, output, description and lookup label, activity should have also the properties: study target and semantically anchored to, as well as parameter set?

File: schemas/research/activity.schema.tpl.json (hunk 10 above)
Comment (apdavison) [reply schemas/research/activity.schema.tpl.json:29]: so it's more like an alias? (i.e. unique)

**Comment by apdavison** [comment]
> > looks good in general, a couple of minor things, and a discussion needed about what to put in input/output
> 
> @apdavison just to be sure: so besides input, output, description and lookup label, activity should have also the properties: study target and semantically anchored to, as well as parameter set?

yes, I think so.

File: schemas/research/activity.schema.tpl.json (hunk 10 above)
Comment (lzehl) [reply schemas/research/activity.schema.tpl.json:29]: it does not have to be unique, but it most likely will be (at least as far as a curator can think of a unique term for the whole KG system)

**Comment by apdavison** [comment]
What do you think about adding fields "startedAtTime", "endedAtTime"? We add these in Computation, and for the ephys extension we would like to add them to subclasses of ProtocolExecution, so maybe having them in Activity, as the base class of all these, is the best place.

**Comment by lzehl** [comment]
> What do you think about adding fields "startedAtTime", "endedAtTime"? We add these in Computation, and for the ephys extension we would like to add them to subclasses of ProtocolExecution, so maybe having them in Activity, as the base class of all these, is the best place.

@apdavison I don't have any objections having them in the activity base class. They should remain optional though, because in many cases these metadata will not be available for a protocol execution.

**Comment by lzehl** [comment]
@apdavison just to be sure: we said now we'll remove the semantically anchored to from the ProtocolExecution / Activity schema correct? And reintroduce it in the TissueSample(Collection) schemas? And in the files?? I think this we need to discuss again (where to reintroduce a user assessed anatomical location of their data). 

Summary of necessary corrections: 
Activity input / output should be reduced to File / FileBundle.
Activity semanticallyAnchoredTo should be removed.
Activity should get startTime / endTime properties (with time and datetime as formats).

**Review by lzehl — APPROVED** [review]

**Comment by lzehl** [comment]
@apdavison I'm merging this PR now, but it will cause some changes to be made in openMINDS_computation. I will go through them and either make a PR there or raise an issue as soon as I can.

**Pull request merged**